src/libmongoc/src/mongoc/mongoc-cluster-aws.c

```diff
@@ -432,6 +488,214 @@ _obtain_creds_from_ecs (_mongoc_aws_credentials_t *creds, bson_error_t *error)
+                           "Response headers: %s",
+                           http_response_headers);
+   }
+
+   if (!bson_iter_init (&iter, response_bson)) {
+      AUTH_ERROR_AND_FAIL ("failed to initialize BSON iterator to response: %s",
+                           http_response_body);
+   }
+
+   // If the response contains "Error", return only the "Error" document.
+   // Do not include http_response_body in subsequent errors to limit risk of
+   // including credentials in error messages.
+   bson_iter_t Error_iter;
+   if (bson_iter_init_find (&Error_iter, response_bson, "Error")) {
+      bson_t Error_bson;
+      if (!_mongoc_iter_document_as_bson (&Error_iter, &Error_bson, error)) {
+         goto fail;
+      }
+      char *Error_json = bson_as_json (&Error_bson, NULL);
+      bson_set_error (
+         error,
+         MONGOC_ERROR_CLIENT,
+         MONGOC_ERROR_CLIENT_AUTHENTICATE,
+         "Response to AssumeRoleWithWebIdentity contains 'Error': %s",
+         Error_json);
+      bson_free (Error_json);
+      goto fail;
+   }
+
+   bson_iter_t Credentials_iter;
+   if (!bson_iter_find_descendant (
+          &iter,
+          "AssumeRoleWithWebIdentityResponse.AssumeRoleWithWebIdentityResult."
+          "Credentials",
+          &iter)) {
+      AUTH_ERROR_AND_FAIL (
+         "did not find "
+         "AssumeRoleWithWebIdentityResponse.AssumeRoleWithWebIdentityResult."
+         "Credentials in response from sts.amazonaws.com.");
+   }
+
+   if (!bson_iter_recurse (&iter, &Credentials_iter)) {
+      AUTH_ERROR_AND_FAIL (
+         "Unable to recurse in Credentials in response from sts.amazonaws.com");
+   }
+
+   iter = Credentials_iter;
+   if (bson_iter_find (&iter, "AccessKeyId") && BSON_ITER_HOLDS_UTF8 (&iter)) {
+      access_key_id = bson_iter_utf8 (&iter, NULL);
+   } else {
+      AUTH_ERROR_AND_FAIL (
+         "did not find AccessKeyId in response from sts.amazonaws.com");
+   }
+
+   iter = Credentials_iter;
+   if (bson_iter_find (&iter, "SecretAccessKey") &&
+       BSON_ITER_HOLDS_UTF8 (&iter)) {
+      secret_access_key = bson_iter_utf8 (&iter, NULL);
+   } else {
+      AUTH_ERROR_AND_FAIL (
+         "did not find SecretAccessKey in response from sts.amazonaws.com");
+   }
+
+   iter = Credentials_iter;
+   if (bson_iter_find (&iter, "SessionToken") && BSON_ITER_HOLDS_UTF8 (&iter)) {
+      session_token = bson_iter_utf8 (&iter, NULL);
+   } else {
+      AUTH_ERROR_AND_FAIL (
+         "did not find SessionToken in response from sts.amazonaws.com");
+   }
+
+   iter = Credentials_iter;
+   if (bson_iter_find (&iter, "Expiration") && BSON_ITER_HOLDS_DOUBLE (&iter)) {
+      // "Expiration" is returned as a double representing the number of seconds
+      // since Unix Epoch. This differs from the ISO-8601 string returned in EC2
+      // and ECS responses.
+      int64_t expiration_ms = ((int64_t) bson_iter_double (&iter)) * 1000;
```

Comment:
Confirming: do we want to apply the `* 1000` in the domain of `int64_t`? Or should it be done in the domain of `double`?
```c
int64_t expiration_ms = (int64_t) (1000.0 * bson_iter_double (&iter));
```

Reply:
The benefit would be avoiding truncation of sub-second digits. As written, the resulting `expiration_ms` will always be an exact multiple of 1,000 since the double value is truncated by the conversion to `int64_t` before multiplication by 1,000:
```c
// 2023-01-01T01:02:03.456Z
const double t = 1672534923.456;
printf("t: %f\n\n", t);

const int64_t ai = (int64_t)t;
const double  bi = 1000.0 * t;
printf("ai: %" PRId64 "\n" "bi: %f\n\n", ai, bi);

const int64_t a = 1000 * ai;
const int64_t b = (int64_t)bi;

printf( "a: %" PRId64 "\n" "b: %" PRId64 "\n", a, b);
```
```
t: 1672534923.456000

ai: 1672534923
bi: 1672534923456.000000

a: 1672534923000
b: 1672534923456
```

Reply:
Oof right, that makes sense. Suggestion has been applied.